{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping accuracy and runtime using strobeAlign"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## StrobeAlign is a new short read aligner designed to be optimum for sequence reads of length 100, 200, 300, 400, 500 bps.\n",
    "## This is highlighted as being the optimal range of base pair alignment for this aligner.\n",
    "The first step of calculating these accuracies for the full genome will be as follow:\n",
    "We have a complete human genome (Genome assembly T2T-CHM13v2.0) of size 3.0 Gb, to cover the full genome with 100 bp reads we will need about 30 million reads (since 3.0 Gb / 100 bp = 30 million reads).\n", 
    "We can optain around the same number of reads using -fcov 1 and -cov 1,250,000 using the read simulator ART (Let's use -cov).\n",
    "\n",
    "#### The direction below can be used to test the runtime and accuracy of single-end and paired-end reads using StrobeAlign for the whole reference genome. \n",
    "\n",
    "#### Further down, there will be a Trial section where I tested smaller reads (24,000 reads) since 30 million reads was crashing my machines.\n",
    "\n",
    "#### Parameter definition \n",
    "\n",
    "-ef  --errfree  indicate to generate the zero sequencing errors SAM file as well the regular one NOTE: the reads in the zero-error SAM file have the same alignment positions as those in the regular SAM file, but have no sequencing errors\n",
    "\n",
    "StrobeAlign can be run either by generating the index on the fly or before the alignment is ran.\n",
    "\n",
    "coverage = number of reads/read pairs to be generated per sequence/amplicon (not be used together with -f/--fcov)\n",
    "The coverage number i.e 1,250,000 for 100 length reads calculated by dividing 30 million reads / 24 sequences in the reference genome.\n",
    "\n",
    "Indels (-maxIndel) =  The parameter used to set the maximum length for indel errors. Indels, insertions and deletions, are a type of sequencing error where bases are erroneously inserted into or deleted from the sequence during the sequencing process. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    import os\n",
    "    import time\n",
    "    from Bio import SeqIO\n",
    "    from Bio.Seq import Seq\n",
    "    import subprocess\n",
    "    import tempfile\n",
    "    import pysam\n",
    "    import matplotlib.pyplot as plt\n",
    "    import itertools\n",
    "except:\n",
    "    print(\"One or more required packages is not installed.\")\n",
    "    print(\"Please verify dependencies and try again.\")\n",
    "    exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lajoyce/Desktop'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reads from the Whole Genome\n",
    "\n",
    "### Reads can be simulated using the ART software. My Trial section has example of command lines I used for these tests. However, below are the numbers that should be used for the overall tests for the whole genome.\n",
    "\n",
    "##### Single-end reads generation command example.\n",
    "\n",
    "art_illumina -ss HS25 -sam -i /home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna -l 100 -c 10000 -ef -rs 10 -o /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/100_long_single_reads_cov\n",
    "\n",
    "##### The -c and -l parameter should be replaced with proper values that represent the experimental environment desired such that:\n",
    "\n",
    "##### -l 100 = -c 1,250,000 (generates 30 million single reads)\n",
    "##### -l 200 = -c 625,000 (generates 15 million single reads)\n",
    "##### -l 300 = -c 416,666.667 (generates 10 million single reads)\n",
    ".\n",
    ".\n",
    ".\n",
    "\n",
    "##### Paired-end reads generation command example.\n",
    "\n",
    "art_illumina -ss HS25 -i GCF_009914755.1_T2T-CHM13v2.0_genomic.fna -p -l 100 -c 1250000 -m 300 -s 20 -ef -o 100_paired_end_reads\n",
    "\n",
    "If -m 300 -s 20 it means the simulated paired-end reads will come from fragments with an average size of 300 pairs, and most of the fragments will fall within 20 base pairs of this average (that is, between 280 and 320 base pairs). This is because in a normal distribution, most values (68%) fall within one standard deviation from the mean.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The download and instructions on how to use StrobeAlign can be found on the link below.\n",
    "\n",
    "https://github.com/ksahlin/strobealign\n",
    "\n",
    "In short, the index matrix used by Strobe can be generated on the flight or by creating the index matrix separately (then add --use-index parameter.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For sanity check we can check that the reads generated in the fastq file is the same as the reads in the error free sam files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the FASTQ file into a dictionary\n",
    "fastq_dict = SeqIO.to_dict(SeqIO.parse(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/single_end_reads/100_long_single_reads_cov.fq\", \"fastq\"))\n",
    "\n",
    "# Open the SAM file\n",
    "samfile = pysam.AlignmentFile(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/single_end_reads/100_long_single_reads_cov_errFree.sam\", \"r\")\n",
    "\n",
    "# Iterate over the reads in the SAM file\n",
    "for read in samfile:\n",
    "    #Get the read indentifier\n",
    "    read_id = read.query_name\n",
    "\n",
    "    #Check if the read identifier exists in the FASTQ dictionary\n",
    "    if read_id in fastq_dict:\n",
    "        # Compare the sequences\n",
    "\n",
    "        if read.query_sequence == str(fastq_dict[read_id].seq):\n",
    "            print(f\"Read {read_id} matches.\")\n",
    "        else:\n",
    "            print(f\"Read {read_id} does not match.\")\n",
    "    else:\n",
    "        print(f\"Read {read_id} not found in FASTQ file.\")\n",
    "\n",
    "# Close the SAM file\n",
    "samfile.close()        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### You can prepare the other reads from the terminal or other methods by changing the length of the reads.\n",
    "\n",
    "##### Running the cell below after generating all the reads with different length should print out the runtimes for all the different reads. The same idea can be followed to get the accuracy for the single and paired end reads too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "reads_names = [\"100_long_single_reads_cov.fq\", \"200_long_single_reads_cov.fq\", \"300_long_single_reads_cov.fq\", \"400_long_single_reads_cov.fq\", \"500_long_single_reads_cov.fq\"]\n",
    "\n",
    "\n",
    "counter = 0 \n",
    "\n",
    "runtimes = []\n",
    "\n",
    "for reads in reads_names:\n",
    "\n",
    "    reads_file = f\"/home/lajoyce/Documents/sequence_alignment/complete_genome/single_end_reads/{reads}\" \n",
    "\n",
    "    reference_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "\n",
    "#output_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/sam_files/100_single_ends_strobe.sam\"\n",
    "\n",
    "\n",
    "\n",
    "# Define the StrobeAlign command \n",
    "# If you want to run strobealign without the predefined index, don't add the \"--use-index\" parameter\n",
    "    strobealign_command = [\"/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign\", \"--use-index\", reference_file, reads_file]\n",
    "\n",
    "# Start the timer\n",
    "\n",
    "    start_time = time.time()\n",
    "\n",
    "    #Run StrobeAlign and discard the output since we may only care about about the runtime\n",
    "\n",
    "    subprocess.run(strobealign_command, stdout=subprocess.DEVNULL, check = True)\n",
    "\n",
    "    # Stop the timer\n",
    "    end_time = time.time()\n",
    "\n",
    "    # Delete the temporary FASTQ file\n",
    "    #os.remove(temp_file_name)\n",
    "\n",
    "    # Calculate and return the runtime \n",
    "    runtime = end_time - start_time\n",
    "\n",
    "    runtimes.append(runtime)\n",
    "\n",
    "    print(f\"Runtime: {runtime} seconds\")\n",
    "\n",
    "    counter += 1\n",
    "\n",
    "    print(counter)\n",
    "\n",
    "print(runtimes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trial (Overall BWA-MEM is more accurate than StrobeAlign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# I had to use smaller reads instead of 30 million to avoid my systems crashing. In general, StrobeAlign uses a lot of RAM to generate the index and obviously running 30 million reads for processing is also memory intensive\n",
    "\n",
    "Adding the errorfree (-ef) parameter generates an additional errorfree sam file that we will use as the ground truth to compare the alignments, especially for accuracy. \n",
    "\n",
    "### Art_illumina command\n",
    "\n",
    "art_illumina -ss HS25 -sam -i /home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna -l 100 -c 10000 -ef -rs 10 -o /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/100_long_single_reads_cov\n",
    "\n",
    "### StrobeAlign command\n",
    "./build/strobealign --use-index /home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/100_long_single_reads_cov.fq > /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/100_single_ends_strobe.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    import os\n",
    "    import time\n",
    "    from Bio import SeqIO\n",
    "    import subprocess\n",
    "    import tempfile\n",
    "    import pysam\n",
    "except:\n",
    "    print(\"One or more required packages is not installed.\")\n",
    "    print(\"Please verify dependencies and try again.\")\n",
    "    exit()\n",
    "\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For sanity check we can check that the reads generated in the fastq file is the same as the reads in the error free sam files and that the standard sam file is different from the errorfree sam files. \n",
    "\n",
    "##### Conclusion: As expected, the reads in each sam files are not the same as the standard sam file includes errors from the Illumina machine that it is simulating while the errorfree sam file does not. \n",
    "\n",
    "##### The reads from the fastq file matches the reads from the standard sam file (these reads are reverse complemented.)\n",
    "\n",
    "##### The code below checks if the reads from the fastq files are the same as the reads from the standard/error free sam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the FASTQ reads into a set \n",
    "\n",
    "fastq_reads = set()\n",
    "for record in SeqIO.parse(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.fq\", \"fastq\"):\n",
    "    fastq_reads.add(str(record.seq))\n",
    "\n",
    "\n",
    "# Open the SAM file and compare each read\n",
    "samfile = pysam.AlignmentFile(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.sam\", \"r\")\n",
    "\n",
    "# Open the errorfree SAM file and compare each read to the fastq file\n",
    "#samfile = pysam.AlignmentFile(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov_errFree.sam\", \"r\")\n",
    "\n",
    "count = 0\n",
    "total_reads = 0\n",
    "for read in samfile:\n",
    "    seq = read.query_sequence\n",
    "    # If the read is reverse complemented (flag 16), reversese complement the sequence before comparison\n",
    "    if read.is_reverse:\n",
    "        seq = str(Seq(seq).reverse_complement())\n",
    "    if seq in fastq_reads:\n",
    "        count += 1\n",
    "    total_reads += 1\n",
    "\n",
    "print(f\"Number of matching reads: {count}\")\n",
    "\n",
    "if count == total_reads:\n",
    "    print(\"All reads matched.\")\n",
    "else:\n",
    "    print(\"Not all reads matched.\")\n",
    "\n",
    "\n",
    "samfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The code below checks if the standard sam file has the same reads as the errorfree sam file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CIGARString comparison\n",
    "\n",
    "def compare_sam_files(file1, file2):\n",
    "    samfile1 = pysam.AlignmentFile(file1, \"r\")\n",
    "    samfile2 = pysam.AlignmentFile(file2, \"r\")\n",
    "\n",
    "    for read1, read2 in zip(samfile1, samfile2):\n",
    "        # Compare essential fields\n",
    "        if (read1.query_name != read2.query_name or read1.reference_start != read2.reference_start or read1.reference_end != read2.reference_end or read1.cigarstring != read2.cigarstring):\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "same = compare_sam_files(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov_errFree.sam\")\n",
    "\n",
    "print(\"SAM files are the same: \", same)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check sequences between both sam files\n",
    "\n",
    "def compare_sam_files_sequences(file1, file2):\n",
    "    samfile1 = pysam.AlignmentFile(file1, \"r\")\n",
    "    samfile2 = pysam.AlignmentFile(file2, \"r\")\n",
    "\n",
    "    for read1, read2 in zip(samfile1, samfile2):\n",
    "        # Reverse complement sequence if needed\n",
    "        seq1 = read1.query_sequence\n",
    "        seq2 = read2.query_sequence\n",
    "        if read1.is_reverse:\n",
    "            seq1 = str(Seq(seq1).reverse_complement())\n",
    "        if read2.is_reverse:\n",
    "            seq2 = str(Seq(seq2).reverse_complement())\n",
    "        \n",
    "        # Compare sequences\n",
    "        if seq1 != seq2:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "same_sequences = compare_sam_files_sequences(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov_errFree.sam\")\n",
    "\n",
    "print(\"SAM files have the same sequences: \", same_sequences)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The code below checks if the starting positions for both sam files are the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile1 = pysam.AlignmentFile(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.sam\", \"r\")\n",
    "samfile2 = pysam.AlignmentFile(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov_errFree.sam\", \"r\")\n",
    "\n",
    "total_reads = 0\n",
    "match_count = 0\n",
    "\n",
    "for read1, read2 in zip(samfile1.fetch(until_eof=True), samfile2.fetch(until_eof=True)):\n",
    "\n",
    "    total_reads += 1\n",
    "    # Compare start positions\n",
    "    if read1.reference_start == read2.reference_start:\n",
    "        match_count += 1\n",
    "\n",
    "print(f\"Number of matching start positions: {count}\")\n",
    "\n",
    "#Check if all the starting positions matched\n",
    "if total_reads == match_count:\n",
    "    print(\"All starting positions matched.\")\n",
    "else:\n",
    "    print(\"Not all starting positions matched.\")\n",
    "\n",
    "samfile1.close()\n",
    "samfile2.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Runtime Single End Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/100_single_end/100_long_single_reads_cov.fq\" \n",
    "\n",
    "reference_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "\n",
    "#output_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/sam_files/100_single_ends_strobe.sam\"\n",
    "\n",
    "\n",
    "\n",
    "# Define the StrobeAlign command \n",
    "# If you want to run strobealign without the predefined index, don't add the \"--use-index\" parameter\n",
    "#strobealign_command = [\"/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign\", \"--use-index\", reference_file, reads_file]\n",
    "\n",
    "# Command to use if you want to keep the output of strobealign\n",
    "#If you want to save the output, add the directory and name of the output file\n",
    "strobealign_command = ' '.join(['/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign', '--use-index', \n",
    "                                '/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna', \n",
    "                                '/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.fq',\n",
    "                                  '> /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/strobealign/single_ends_strobe.sam'])\n",
    "\n",
    "# Start the timer\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "#try:\n",
    "#Run StrobeAlign and do not discard the output\n",
    "subprocess.run(strobealign_command, check=True, shell=True)\n",
    "\n",
    "#except subprocess.CalledProcessError as e:\n",
    "#    print(\"Error:\")\n",
    " #   print(e.stderr)\n",
    "\n",
    "#Run StrobeAlign and discard the output since we may only care about about the runtime\n",
    "\n",
    "#subprocess.run(strobealign_command, stdout=subprocess.DEVNULL, check = True)\n",
    "\n",
    "# Stop the timer\n",
    "end_time = time.time()\n",
    "\n",
    "#Delete the temporary FASTQ file\n",
    "#os.remove(temp_file_name)\n",
    "\n",
    "# Calculate and return the runtime \n",
    "runtime = end_time - start_time\n",
    "\n",
    "print(f\"Runtime: {runtime} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run time for all the reads (single-end reads)\n",
    "\n",
    "### The idea is to generate reads with different length and then test it all together and plot the results. This can be done for paired-end reads etc..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "reads_names = [\"100_long_single_reads_cov.fq\", \"200_long_single_reads_cov.fq\", \"300_long_single_reads_cov.fq\", \"400_long_single_reads_cov.fq\", \"500_long_single_reads_cov.fq\"]\n",
    "\n",
    "\n",
    "counter = 0 \n",
    "\n",
    "runtimes = []\n",
    "\n",
    "for reads in reads_names:\n",
    "\n",
    "    reads_file = f\"/home/lajoyce/Documents/sequence_alignment/complete_genome/single_end_reads/{reads}\" \n",
    "\n",
    "    reference_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "\n",
    "#output_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/sam_files/100_single_ends_strobe.sam\"\n",
    "\n",
    "\n",
    "\n",
    "# Define the StrobeAlign command \n",
    "# If you want to run strobealign without the predefined index, don't add the \"--use-index\" parameter\n",
    "    strobealign_command = [\"/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign\", \"--use-index\", reference_file, reads_file]\n",
    "\n",
    "# Start the timer\n",
    "\n",
    "    start_time = time.time()\n",
    "\n",
    "    #Run StrobeAlign and discard the output since we may only care about about the runtime\n",
    "\n",
    "    subprocess.run(strobealign_command, stdout=subprocess.DEVNULL, check = True)\n",
    "\n",
    "    # Stop the timer\n",
    "    end_time = time.time()\n",
    "\n",
    "    # Delete the temporary FASTQ file\n",
    "    #os.remove(temp_file_name)\n",
    "\n",
    "    # Calculate and return the runtime \n",
    "    runtime = end_time - start_time\n",
    "\n",
    "    runtimes.append(runtime)\n",
    "\n",
    "    print(f\"Runtime: {runtime} seconds\")\n",
    "\n",
    "    counter += 1\n",
    "\n",
    "    print(counter)\n",
    "\n",
    "print(runtimes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Accuracy & Multi-mapping reading\n",
    "\n",
    "### Comparing starting positions only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_accuracy_and_multimapping(ground_truth_sam, test_sam):\n",
    "    # Initialize variables\n",
    "    total_reads = 0\n",
    "    correct_alignments = 0\n",
    "    multimapped_reads = 0\n",
    "\n",
    "    with pysam.AlignmentFile(ground_truth_sam, \"r\") as gt_sam, pysam.AlignmentFile(test_sam, \"r\") as t_sam:\n",
    "        for truth, test in zip(gt_sam, t_sam):\n",
    "            # Increase total reads\n",
    "            total_reads += 1\n",
    "            \n",
    "            # Check if alignments are the same\n",
    "            if truth.reference_start == test.reference_start:\n",
    "                correct_alignments += 1\n",
    "\n",
    "            # Check if read is multi-mapped\n",
    "            if test.has_tag(\"XA\"):\n",
    "                multimapped_reads += 1\n",
    "                \n",
    "    # If no reads found, print a message and return\n",
    "    if total_reads == 0:\n",
    "        print(\"No reads found in the SAM files.\")\n",
    "        return 0, 0\n",
    "    \n",
    "    # Calculate the accuracy\n",
    "    accuracy = correct_alignments / total_reads\n",
    "    multimapping_percentage = multimapped_reads / total_reads\n",
    "    print(f\"The number of correct alignments: {correct_alignments}\")\n",
    "\n",
    "    return accuracy, multimapping_percentage\n",
    "\n",
    "accuracy, multimapping_percentage = calculate_accuracy_and_multimapping(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/strobealign/single_ends_strobe.sam\")\n",
    "\n",
    "print(f\"Alignment accuracy: {accuracy * 100}%\")\n",
    "print(f\"Percentage of multi-mapped reads: {multimapping_percentage * 100}%\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check for starting and ending positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_accuracy_and_multimapping(ground_truth_sam, test_sam):\n",
    "    # Initialize variables\n",
    "    total_reads = 0\n",
    "    correct_alignments = 0\n",
    "    multimapped_reads = 0\n",
    "\n",
    "    with pysam.AlignmentFile(ground_truth_sam, \"r\") as gt_sam, pysam.AlignmentFile(test_sam, \"r\") as t_sam:\n",
    "        for truth, test in zip(gt_sam, t_sam):\n",
    "            # Increase total reads\n",
    "            total_reads += 1\n",
    "\n",
    "            # Compare start and inferred end positions\n",
    "            if truth.reference_start == test.reference_start and truth.infer_read_length() == test.infer_read_length():\n",
    "                correct_alignments += 1\n",
    "\n",
    "            # Check if read is multi-mapped\n",
    "            if test.has_tag(\"XA\"):\n",
    "                multimapped_reads += 1\n",
    "\n",
    "    # If no reads found, print a message and return\n",
    "    if total_reads == 0:\n",
    "        print(\"No reads found in the SAM files.\")\n",
    "        return 0, 0\n",
    "\n",
    "    # Calculate the accuracy\n",
    "    accuracy = correct_alignments / total_reads\n",
    "    multimapping_percentage = multimapped_reads / total_reads\n",
    "\n",
    "    return accuracy, multimapping_percentage\n",
    "\n",
    "# Use the function\n",
    "accuracy, multimapping_percentage = calculate_accuracy_and_multimapping(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/strobealign/single_ends_strobe.sam\")\n",
    "\n",
    "print(f\"Alignment accuracy: {accuracy * 100}%\")\n",
    "print(f\"Percentage of multi-mapped reads: {multimapping_percentage * 100}%\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paired-end reads \n",
    "\n",
    "### ART can generate paired-end reads.\n",
    "\n",
    "##### The reads I generated for my trial tests was done by the command below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters\n",
    "reference_genome = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "output_filename = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads\"\n",
    "read_length = 100  # The desired read length\n",
    "coverage = 1000  # The desired coverage\n",
    "insert_size = 300  # The desired mean insert size\n",
    "insert_std_dev = 20  # The desired standard deviation for the insert size\n",
    "\n",
    "# Command\n",
    "command = f\"art_illumina -ss HS25 -i {reference_genome} -p -l {read_length} -c {coverage} -m {insert_size} -s {insert_std_dev} -na -ef -o {output_filename}\"\n",
    "\n",
    "# Run the command\n",
    "process = subprocess.Popen(command, shell=True)\n",
    "\n",
    "process.wait()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The separated FASTQ files can be conbined into interleaved fastq files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def interleave(iter1, iter2):\n",
    "    # Interleave the reads from two iterators\n",
    "    return itertools.chain.from_iterable(itertools.zip_longest(iter1, iter2))\n",
    "\n",
    "file1 = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads1.fq\"\n",
    "file2 = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads2.fq\"\n",
    "outfile = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/interleaved_reads.fq\"\n",
    "\n",
    "read1 = SeqIO.parse(file1, \"fastq\")\n",
    "read2 = SeqIO.parse(file2, \"fastq\")\n",
    "\n",
    "interleaved = interleave(read1, read2)\n",
    "\n",
    "count = SeqIO.write(interleaved, outfile, \"fastq\")\n",
    "print(\"Interleaved %i reads\" % count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Paired end reads (interleaved) StrobeAlign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/100_single_end/100_long_single_reads_cov.fq\" \n",
    "\n",
    "reference_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "\n",
    "#output_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/sam_files/100_single_ends_strobe.sam\"\n",
    "\n",
    "\n",
    "\n",
    "# Define the StrobeAlign command \n",
    "# If you want to run strobealign without the predefined index, don't add the \"--use-index\" parameter\n",
    "#strobealign_command = [\"/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign\", \"--use-index\", reference_file, reads_file]\n",
    "\n",
    "# Command to use if you want to keep the output of strobealign\n",
    "#If you want to save the output, add the directory and name of the output file\n",
    "\n",
    "# To report secondary alignments, set parameter -N [INT] for a maximum of [INT] secondary alignments.\n",
    "strobealign_command = ' '.join(['/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign', '--use-index', \n",
    "                                '/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna', \n",
    "                                '/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/interleaved_reads.fq',\n",
    "                                  '> /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/strobealign/paired_end_reads/100_paired_ends_strobe.sam'])\n",
    "\n",
    "# Start the timer\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "#try:\n",
    "#Run StrobeAlign and do not discard the output\n",
    "subprocess.run(strobealign_command, check=True, shell=True)\n",
    "\n",
    "#except subprocess.CalledProcessError as e:\n",
    "#    print(\"Error:\")\n",
    " #   print(e.stderr)\n",
    "\n",
    "#Run StrobeAlign and discard the output since we may only care about about the runtime\n",
    "\n",
    "#subprocess.run(strobealign_command, stdout=subprocess.DEVNULL, check = True)\n",
    "\n",
    "# Stop the timer\n",
    "end_time = time.time()\n",
    "\n",
    "#Delete the temporary FASTQ file\n",
    "#os.remove(temp_file_name)\n",
    "\n",
    "# Calculate and return the runtime \n",
    "runtime = end_time - start_time\n",
    "\n",
    "print(f\"Runtime: {runtime} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accuracy Check Starting Position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_accuracy_and_multimapping(ground_truth_sam, test_sam):\n",
    "    # Initialize variables\n",
    "    total_reads = 0\n",
    "    correct_alignments = 0\n",
    "    multimapped_reads = 0\n",
    "\n",
    "    with pysam.AlignmentFile(ground_truth_sam, \"r\") as gt_sam, pysam.AlignmentFile(test_sam, \"r\") as t_sam:\n",
    "        for truth, test in zip(gt_sam, t_sam):\n",
    "            # Increase total reads\n",
    "            total_reads += 1\n",
    "            \n",
    "            # Check if alignments are the same\n",
    "            if truth.reference_start == test.reference_start:\n",
    "                correct_alignments += 1\n",
    "\n",
    "            # Check if read is multi-mapped\n",
    "            if test.has_tag(\"XA\"):\n",
    "                multimapped_reads += 1\n",
    "                \n",
    "    # If no reads found, print a message and return\n",
    "    if total_reads == 0:\n",
    "        print(\"No reads found in the SAM files.\")\n",
    "        return 0, 0\n",
    "    \n",
    "    # Calculate the accuracy\n",
    "    accuracy = correct_alignments / total_reads\n",
    "    multimapping_percentage = multimapped_reads / total_reads\n",
    "    print(f\"The number of correct alignments: {correct_alignments}\")\n",
    "\n",
    "    return accuracy, multimapping_percentage\n",
    "\n",
    "accuracy, multimapping_percentage = calculate_accuracy_and_multimapping(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/strobealign/paired_end_reads/100_paired_ends_strobe.sam\")\n",
    "\n",
    "print(f\"Alignment accuracy: {accuracy * 100}%\")\n",
    "print(f\"Percentage of multi-mapped reads: {multimapping_percentage * 100}%\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Accuracy checking for starting and ending positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_accuracy_and_multimapping(ground_truth_sam, test_sam):\n",
    "    # Initialize variables\n",
    "    total_reads = 0\n",
    "    correct_alignments = 0\n",
    "    multimapped_reads = 0\n",
    "\n",
    "    with pysam.AlignmentFile(ground_truth_sam, \"r\") as gt_sam, pysam.AlignmentFile(test_sam, \"r\") as t_sam:\n",
    "        for truth, test in zip(gt_sam, t_sam):\n",
    "            # Increase total reads\n",
    "            total_reads += 1\n",
    "\n",
    "            # Compare start and inferred end positions\n",
    "            if truth.reference_start == test.reference_start and truth.infer_read_length() == test.infer_read_length():\n",
    "                correct_alignments += 1\n",
    "\n",
    "            # Check if read is multi-mapped\n",
    "            if test.has_tag(\"XA\"):\n",
    "                multimapped_reads += 1\n",
    "\n",
    "    # If no reads found, print a message and return\n",
    "    if total_reads == 0:\n",
    "        print(\"No reads found in the SAM files.\")\n",
    "        return 0, 0\n",
    "\n",
    "    # Calculate the accuracy\n",
    "    accuracy = correct_alignments / total_reads\n",
    "    multimapping_percentage = multimapped_reads / total_reads\n",
    "\n",
    "    return accuracy, multimapping_percentage\n",
    "\n",
    "# Use the function\n",
    "accuracy, multimapping_percentage = calculate_accuracy_and_multimapping(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/strobealign/paired_end_reads/100_paired_ends_strobe.sam\")\n",
    "\n",
    "print(f\"Alignment accuracy: {accuracy * 100}%\")\n",
    "print(f\"Percentage of multi-mapped reads: {multimapping_percentage * 100}%\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusion: Accuracy here should be the same as simply checking the starting positions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BWA-MEM Double Check \n",
    "\n",
    "### BWA-MEM2 gave me errors when I tried to generate the index for the reference genome with it.\n",
    "\n",
    "### We are going to use BWA-MEM instead then."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/bwa-mem2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test for Runtime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/100_single_end/100_long_single_reads_cov.fq\" \n",
    "\n",
    "reference_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "\n",
    "#output_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/sam_files/100_single_ends_strobe.sam\"\n",
    "\n",
    "\n",
    "# Define the StrobeAlign command \n",
    "# If you want to run strobealign without the predefined index, don't add the \"--use-index\" parameter\n",
    "#strobealign_command = [\"/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign\", \"--use-index\", reference_file, reads_file]\n",
    "\n",
    "# Command to use if you want to keep the output of strobealign\n",
    "#If you want to save the output, add the directory and name of the output file\n",
    "bwamem2_command = ' '.join([\"/home/lajoyce/Documents/sequence_alignment/bwa/bwa\",\n",
    "                            \"mem\",\n",
    "                            \"GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\",\n",
    "                            \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.fq\",\n",
    "                             \"> /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/bwa-mem2/100_long_single_end_reads.sam\"])\n",
    "                            \n",
    "# Start the timer\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "#try:\n",
    "#Run StrobeAlign and do not discard the output\n",
    "subprocess.run(bwamem2_command, check=True, shell=True)\n",
    "\n",
    "#except subprocess.CalledProcessError as e:\n",
    "#    print(\"Error:\")\n",
    " #   print(e.stderr)\n",
    "\n",
    "#Run StrobeAlign and discard the output since we may only care about about the runtime\n",
    "\n",
    "#subprocess.run(strobealign_command, stdout=subprocess.DEVNULL, check = True)\n",
    "\n",
    "# Stop the timer\n",
    "end_time = time.time()\n",
    "\n",
    "#Delete the temporary FASTQ file\n",
    "#os.remove(temp_file_name)\n",
    "\n",
    "# Calculate and return the runtime \n",
    "runtime = end_time - start_time\n",
    "\n",
    "print(f\"Runtime: {runtime} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare BWA-MEM SAM file to the Ground Truth File for Accuracy\n",
    "\n",
    "#### Here we compare starting positions only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_accuracy_and_multimapping(ground_truth_sam, test_sam):\n",
    "    # Initialize variables\n",
    "    total_reads = 0\n",
    "    correct_alignments = 0\n",
    "    multimapped_reads = 0\n",
    "\n",
    "    with pysam.AlignmentFile(ground_truth_sam, \"r\") as gt_sam, pysam.AlignmentFile(test_sam, \"r\") as t_sam:\n",
    "        for truth, test in zip(gt_sam, t_sam):\n",
    "            # Increase total reads\n",
    "            total_reads += 1\n",
    "            \n",
    "            # Check if alignments are the same\n",
    "            if truth.reference_start == test.reference_start:\n",
    "                correct_alignments += 1\n",
    "\n",
    "            # Check if read is multi-mapped\n",
    "            if test.has_tag(\"XA\"):\n",
    "                multimapped_reads += 1\n",
    "                \n",
    "    # If no reads found, print a message and return\n",
    "    if total_reads == 0:\n",
    "        print(\"No reads found in the SAM files.\")\n",
    "        return 0, 0\n",
    "    \n",
    "    # Calculate the accuracy\n",
    "    accuracy = correct_alignments / total_reads\n",
    "    multimapping_percentage = multimapped_reads / total_reads\n",
    "    print(f\"The number of correct alignments: {correct_alignments}\")\n",
    "\n",
    "    return accuracy, multimapping_percentage\n",
    "\n",
    "accuracy, multimapping_percentage = calculate_accuracy_and_multimapping(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/100_long_single_reads_cov.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/bwa-mem2/100_long_sing_end_reads_bwa-mem.sam\")\n",
    "\n",
    "print(f\"Alignment accuracy: {accuracy * 100}%\")\n",
    "print(f\"Percentage of multi-mapped reads: {multimapping_percentage * 100}%\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paired-end Reads Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/100_single_end/100_long_single_reads_cov.fq\" \n",
    "\n",
    "reference_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "\n",
    "#output_file = \"/home/lajoyce/Documents/sequence_alignment/complete_genome/sam_files/100_single_ends_strobe.sam\"\n",
    "\n",
    "\n",
    "# Define the StrobeAlign command \n",
    "# If you want to run strobealign without the predefined index, don't add the \"--use-index\" parameter\n",
    "#strobealign_command = [\"/home/lajoyce/Documents/sequence_alignment/strobealign/build/strobealign\", \"--use-index\", reference_file, reads_file]\n",
    "\n",
    "# Command to use if you want to keep the output of strobealign\n",
    "#If you want to save the output, add the directory and name of the output file\n",
    "bwamem2_command = ' '.join([\"/home/lajoyce/Documents/sequence_alignment/bwa/bwa\",\n",
    "                            \"mem\",\n",
    "                            \"GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\",\n",
    "                            \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads1.fq\",\n",
    "                            \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads2.fq\"\n",
    "                             \"> /home/lajoyce/Documents/sequence_alignment/complete_genome/trial/bwa-mem2/paired_end_reads.sam\"])\n",
    "                            \n",
    "# Start the timer\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "#try:\n",
    "#Run StrobeAlign and do not discard the output\n",
    "subprocess.run(bwamem2_command, check=True, shell=True)\n",
    "\n",
    "#except subprocess.CalledProcessError as e:\n",
    "#    print(\"Error:\")\n",
    " #   print(e.stderr)\n",
    "\n",
    "#Run StrobeAlign and discard the output since we may only care about about the runtime\n",
    "\n",
    "#subprocess.run(strobealign_command, stdout=subprocess.DEVNULL, check = True)\n",
    "\n",
    "# Stop the timer\n",
    "end_time = time.time()\n",
    "\n",
    "#Delete the temporary FASTQ file\n",
    "#os.remove(temp_file_name)\n",
    "\n",
    "# Calculate and return the runtime \n",
    "runtime = end_time - start_time\n",
    "\n",
    "print(f\"Runtime: {runtime} seconds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_accuracy_and_multimapping(ground_truth_sam, test_sam):\n",
    "    # Initialize variables\n",
    "    total_reads = 0\n",
    "    correct_alignments = 0\n",
    "    multimapped_reads = 0\n",
    "\n",
    "    with pysam.AlignmentFile(ground_truth_sam, \"r\") as gt_sam, pysam.AlignmentFile(test_sam, \"r\") as t_sam:\n",
    "        for truth, test in zip(gt_sam, t_sam):\n",
    "            # Increase total reads\n",
    "            total_reads += 1\n",
    "            \n",
    "            # Check if alignments are the same\n",
    "            if truth.reference_start == test.reference_start:\n",
    "                correct_alignments += 1\n",
    "\n",
    "            # Check if read is multi-mapped\n",
    "            if test.has_tag(\"XA\"):\n",
    "                multimapped_reads += 1\n",
    "                \n",
    "    # If no reads found, print a message and return\n",
    "    if total_reads == 0:\n",
    "        print(\"No reads found in the SAM files.\")\n",
    "        return 0, 0\n",
    "    \n",
    "    # Calculate the accuracy\n",
    "    accuracy = correct_alignments / total_reads\n",
    "    multimapping_percentage = multimapped_reads / total_reads\n",
    "    print(f\"The number of correct alignments: {correct_alignments}\")\n",
    "\n",
    "    return accuracy, multimapping_percentage\n",
    "\n",
    "accuracy, multimapping_percentage = calculate_accuracy_and_multimapping(\"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/reads_simulated/paired_end_reads/100_paired_end_reads.sam\", \"/home/lajoyce/Documents/sequence_alignment/complete_genome/trial/bwa-mem2/paired_end_reads.sam\")\n",
    "\n",
    "print(f\"Alignment accuracy: {accuracy * 100}%\")\n",
    "print(f\"Percentage of multi-mapped reads: {multimapping_percentage * 100}%\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
